granule.go

```diff
@@ -38,52 +36,49 @@ type GranuleMetadata struct {
 	pruned atomic.Uint64
 }
 
-func NewGranule(granulesCreated prometheus.Counter, tableConfig *TableConfig, firstPart *parts.Part) (*Granule, error) {
+func NewGranule(granulesCreated prometheus.Counter, tableConfig *TableConfig, prts ...*parts.Part) (*Granule, error) {
 	g := &Granule{
 		granulesCreated: granulesCreated,
 		parts:           parts.NewList(&atomic.Pointer[parts.Node]{}, parts.None),
 		tableConfig:     tableConfig,
-
-		metadata: GranuleMetadata{
-			least: atomic.Pointer[dynparquet.DynamicRow]{},
-		},
+		metadata:        GranuleMetadata{},
 	}
 
-	// Find the "smallest" row
-	if firstPart != nil {
-		g.metadata.size.Store(uint64(firstPart.Buf.ParquetFile().Size()))
-		g.parts.Prepend(firstPart)
-		least, err := firstPart.Least()
-		if err != nil {
+	for _, p := range prts {
+		if err := g.addPart(p); err != nil {
 			return nil, err
 		}
-		g.metadata.least.Store(least)
 	}
 
 	granulesCreated.Inc()
 	return g, nil
 }
 
-func (g *Granule) addPart(p *parts.Part, r *dynparquet.DynamicRow) (uint64, error) {
-	rows := p.Buf.NumRows()
-	if rows == 0 {
-		return g.metadata.size.Load(), nil
+func (g *Granule) addPart(p *parts.Part) error {
+	if p.NumRows() == 0 {
+		return nil
 	}
-	node := g.parts.Prepend(p)
 
-	newSize := g.metadata.size.Add(uint64(p.Buf.ParquetFile().Size()))
+	r, err := p.Least()
+	if err != nil {
+		return err
+	}
 
-	for {
-		least := g.metadata.least.Load()
-		if least == nil || g.tableConfig.schema.RowLessThan(r, (*dynparquet.DynamicRow)(least)) {
-			if g.metadata.least.CompareAndSwap(least, r) {
-				break
-			}
-		} else {
-			break
-		}
+	_ = g.parts.Prepend(p)
+	g.metadata.size.Add(uint64(p.Size()))
+
+	if g.metadata.least == nil || g.tableConfig.schema.RowLessThan(r, g.metadata.least) {
+		g.metadata.least = r
```

Comment:
I'm not sure I understand, I thought this commit made least immutable but here you're overwriting with a new part's least row?

Reply:
So this _only_ happens at `NewGranule` creation, which you can initialize a granule with many parts (we do this in compaction). During that initialization we find and set the least row to use as a sorting key. After creation that value is no longer able to be updated.